AGN Triggering

forrestglines opened the pull request:
Adds multiphysics modules necessary for evolving galaxy clusters with triggered AGN feedback. It implements the three triggering mechanisms from [Meece 2017](https://doi.org/10.3847/1538-4357/aa6fb1), cold gas triggering, boosted Bondi accretion, and Booth Schaye accretion. It both calculates the AGN feedback to inject from these mechanisms and removes mass from the simulation at a fixed velocity and temperature (so that momentum and specific gas energy is changed). I've also modified the thermal AGN feedback to inject mass at a fixed velocity (so that momentum is changed) and magnetic AGN feedback mass injection at a fixed velocity and temperature (so that momentum and specific gas energy is changed). These implementation details/choices are intended to follow Meece 2017 as closely as possible, although they could be revisited.

 The separate components added are:

- Calculation of AGN triggering from Cold Gas
- Calculation of AGN triggering from Boosted Bondi
- Calculation of AGN triggering from Booth-Schaye
- Gas removal (ie. accretion) of triggering gas at fixed velocity and temperature
- Mass injection at fixed velocity  with Thermal AGN feedback
- Mass injection at fixed velocity, fixed temperature mass injection with Magnetic AGN feedback

Each of these separate components have been tested individually in regression tests to function as expected (not to say they could still have bugs). I've also ran a larger test with cold gas accretion and all three kinds of AGN feedback (thermal, kinetic jet, and magnetic feedback) working at once across 4 gpus on a node with multiple restarts -- everything seems to work as of commit 8b789409ffe37d60164515e0300c9b2afa5b3709. This test is probably sufficient for preliminary science runs at larger scales, that would be my next step if I had the time.

I think this MR is ready for review and to be merged. There are a few outstanding questions about implementation choices for either reviewers or future merge requests. However, since the changes in this MR are exclusive to running galaxy clusters, I'd argue for merging these changes in after review.

My outstanding questions and things for reviewers to double check are:

- Calculation of AGN triggering is implementing via hacking the Parthenon history dump feature, which allows doing reductions across the simulation. These happen at the end of an evolution step whereas it might be more correct to calculate triggering before the step. Hooks for such reductions might have already been added to Parthenon.
- Are the primitive variables used for calculating AGN triggering at the correct time stamp or are they from the previous timestep?
- Double check that the AGN triggering calculated/used after a restart is the same if the simulation had continued through that step without stopping and restarting.
- Is the implementation of boosted Bondi/Booth-Schaye correct as to how density, velocity, and sound-speed averaged over the accretion volume?
- Is the mass injection at fixed velocity with thermal feedback and at fixed velocity and temperature with magnetic feedback both implemented correctly and appropriate for the AGN-physics model we want?

On a related note for reviewers, if anything does seem broken/not working, I'd suggest checking whether the input file is correct -- that things are named correctly, that they match the inputs read by the code, and that the inputs are saved correctly in the code.

There are a three things left to implement for galaxy cluster simulations but we could move forward with preliminary simulations. These features are:

- Initial sinusoidal perturbations to density/velocity/magnetic fields
- Supernova feedback following [Prasad 2022](https://doi.org/10.3847/1538-4357/abc33c) / [Voit 2015](https://ui.adsabs.harvard.edu/link_gateway/2017ApJ...845...80V/doi:10.3847/1538-4357/aa7d04)
- AMR (adaptive, not just static) on density or cooling rate

I have a repo on gitlab for creating initial conditions for galaxy clusters, making movies, and plotting statistics. [magnetized-clusters](https://gitlab.com/forrestglines/magnetized-clusters). You need Python 3 and Jupyter notebook with both [unyt](https://unyt.readthedocs.io/en/stable/) and the Parthenon-enabled yt [from my `parthenon-frontend` branch on my github repo](https://github.com/forrestglines/yt/tree/parthenon-frontend) installed.

A few useful files and notebooks from my `magnetized-clusters` repo are:

[Generate AthenaPK Input.ipynb](https://gitlab.com/forrestglines/magnetized-clusters/-/blob/main/notebooks/cluster_testing/Generate%20AthenaPK%20Input.ipynb): Generate input files for galaxy clusters. Since I've designed the input files to use purely code units (except to define code units), it's helpful to use `unyt` to manage the code units that go into the input file.

[Slice Movies.ipynb](https://gitlab.com/forrestglines/magnetized-clusters/-/blob/main/notebooks/cluster_testing/Slice%20Movies.ipynb): Make movies from galaxy cluster simulations. Does the heavy lifting for automatically generating boundaries for the color/field axes on plots so that the axes aren't constantly changing throughout the movie but the axis bounds also don't have to be explicitly set.


[Statistics.ipynb](https://gitlab.com/forrestglines/magnetized-clusters/-/blob/main/notebooks/cluster_testing/Statistics.ipynb): WIP notebook to read statistics over time from a simulation. E.g. total mass/cooling over time, mins and maxes, cold gas mass, etc. Data reading is fleshed out, plots/figures are WIP

[medium.precessing_all_mech_jet.input](https://gitlab.com/forrestglines/magnetized-clusters/-/blob/main/inputs/cluster_testing/misc_tests/medium.precessing_all_mechs_jet.input): A working, with restarts, half-handcrafted, half-generated input file including a cold-gas triggered precessing AGN jet with thermal, kinetic, and magnetic AGN feedback. Would be a good starting point for a full simulation.

forrestglines commented:
@pgrete This PR was ready for review but after merging with main many of the cluster tests are failing due to negative densities in tests that should be running with a static hydro, source terms only. It seems that `riemann = none` is no longer skipping the hydro step like it used to. Can you take a look at it?

Also, I don't seem to have any write access to this repo, I can't push anything up

pgrete commented:
> @pgrete This PR was ready for review but after merging with main many of the cluster tests are failing due to negative densities in tests that should be running with a static hydro, source terms only. It seems that `riemann = none` is no longer skipping the hydro step like it used to. Can you take a look at it?
> 
> Also, I don't seem to have any write access to this repo, I can't push anything up

I'll have a look.

pgrete commented:
I just merged `main` in and both on the host and on Cuda device all cluster tests pass:
```
The following tests passed:
	regression_test:cluster_hse
	regression_test:cluster_tabular_cooling
	regression_test:cluster_magnetic_tower
	regression_test:cluster_hydro_agn_feedback
	regression_test:cluster_agn_triggering

100% tests passed, 0 tests failed out of 5
```

Did you update the submodule when you tried and/or can you confirm that the current version now also works on your end?

forrestglines commented:
I think I had messed up the merge with `hydro.cpp`. It's working on my end now

forrestglines commented:
Hmmm, two of these tests are sometimes failing on the A40 nodes on Delta GPU while still working on the A100. Even worse, one of them is passing when you run it separately. 

Here's my setup:
```
glines@dt-login02:/u/glines/code/athenapk-project/builds/ 
$ module list

Currently Loaded Modules:
  1) cue-login-env/1.0   2) gcc/11.2.0   3) modtree/gpu   4) default   5) openmpi/4.1.4   6) cuda/11.7.0   7) hdf5/1.12.2   8) cudnn/8.4.1.50   9) anaconda3_gpu/4.13.0
```

Here's running the different tests.
```
glines@gpub080:/u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86/ 
$ ctest -R cluster
Test project /u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86
    Start  6: regression_test:cluster_hse
1/5 Test  #6: regression_test:cluster_hse ..................   Passed   28.09 sec
    Start  7: regression_test:cluster_tabular_cooling
2/5 Test  #7: regression_test:cluster_tabular_cooling ......   Passed   19.75 sec
    Start 11: regression_test:cluster_magnetic_tower
3/5 Test #11: regression_test:cluster_magnetic_tower .......***Failed    8.35 sec
    Start 12: regression_test:cluster_hydro_agn_feedback
4/5 Test #12: regression_test:cluster_hydro_agn_feedback ...***Failed   16.70 sec
    Start 13: regression_test:cluster_agn_triggering
5/5 Test #13: regression_test:cluster_agn_triggering .......   Passed   21.19 sec

60% tests passed, 2 tests failed out of 5

Label Time Summary:
convergence    =  94.08 sec*proc (5 tests)
regression     =  94.08 sec*proc (5 tests)
serial         =  94.08 sec*proc (5 tests)

Total Test time (real) =  94.10 sec

The following tests FAILED:
         11 - regression_test:cluster_magnetic_tower (Failed)
         12 - regression_test:cluster_hydro_agn_feedback (Failed)
Errors while running CTest
glines@gpub080:/u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86/ 
$ ctest -R cluster_magnetic_tower
Test project /u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86
    Start 11: regression_test:cluster_magnetic_tower
1/1 Test #11: regression_test:cluster_magnetic_tower ...***Failed    8.82 sec

0% tests passed, 1 tests failed out of 1

Label Time Summary:
convergence    =   8.82 sec*proc (1 test)
regression     =   8.82 sec*proc (1 test)
serial         =   8.82 sec*proc (1 test)

Total Test time (real) =   8.84 sec

The following tests FAILED:
         11 - regression_test:cluster_magnetic_tower (Failed)
Errors while running CTest
glines@gpub080:/u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86/ 
$ ctest -R cluster_hydro_agn
Test project /u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86
    Start 12: regression_test:cluster_hydro_agn_feedback
1/1 Test #12: regression_test:cluster_hydro_agn_feedback ...   Passed   34.86 sec

100% tests passed, 0 tests failed out of 1

Label Time Summary:
convergence    =  34.86 sec*proc (1 test)
regression     =  34.86 sec*proc (1 test)
serial         =  34.86 sec*proc (1 test)

Total Test time (real) =  34.87 sec

glines@gpub075:/u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86/ 
$ ctest -R cluster
Test project /u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86
    Start  6: regression_test:cluster_hse
1/5 Test  #6: regression_test:cluster_hse ..................   Passed   45.51 sec
    Start  7: regression_test:cluster_tabular_cooling
2/5 Test  #7: regression_test:cluster_tabular_cooling ......   Passed   22.73 sec
    Start 11: regression_test:cluster_magnetic_tower
3/5 Test #11: regression_test:cluster_magnetic_tower .......***Failed    7.60 sec
    Start 12: regression_test:cluster_hydro_agn_feedback
4/5 Test #12: regression_test:cluster_hydro_agn_feedback ...   Passed   33.36 sec
    Start 13: regression_test:cluster_agn_triggering
5/5 Test #13: regression_test:cluster_agn_triggering .......   Passed   20.51 sec

80% tests passed, 1 tests failed out of 5

Label Time Summary:
convergence    = 129.71 sec*proc (5 tests)
regression     = 129.71 sec*proc (5 tests)
serial         = 129.71 sec*proc (5 tests)

Total Test time (real) = 129.83 sec

The following tests FAILED:
         11 - regression_test:cluster_magnetic_tower (Failed)
Errors while running CTest
glines@gpub075:/u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86/ 
$ ctest -R cluster
Test project /u/glines/code/athenapk-project/builds/cuda-RelWithDebInfo-ampere86
    Start  6: regression_test:cluster_hse
1/5 Test  #6: regression_test:cluster_hse ..................   Passed   27.77 sec
    Start  7: regression_test:cluster_tabular_cooling
2/5 Test  #7: regression_test:cluster_tabular_cooling ......   Passed   17.43 sec
    Start 11: regression_test:cluster_magnetic_tower
3/5 Test #11: regression_test:cluster_magnetic_tower .......***Failed   16.83 sec
    Start 12: regression_test:cluster_hydro_agn_feedback
4/5 Test #12: regression_test:cluster_hydro_agn_feedback ...   Passed   33.21 sec
    Start 13: regression_test:cluster_agn_triggering
5/5 Test #13: regression_test:cluster_agn_triggering .......***Failed    7.04 sec

60% tests passed, 2 tests failed out of 5

Label Time Summary:
convergence    = 102.29 sec*proc (5 tests)
regression     = 102.29 sec*proc (5 tests)
serial         = 102.29 sec*proc (5 tests)

Total Test time (real) = 102.31 sec

The following tests FAILED:
         11 - regression_test:cluster_magnetic_tower (Failed)
         13 - regression_test:cluster_agn_triggering (Failed)
Errors while running CTest

```
Here's the LastTest.log: https://pastebin.com/irFD2jrm

Looks like negative density errors. Note that for these tests that are failing the hydro should be passive with source terms only.

Given that there seem to be issues running on Delta this week, maybe we should wait until next week to see if it resolves itself?

forrestglines commented:
Small update on the Delta A40s - these tests pass just fine under a DEBUG build.

pgrete commented:
Just do double check: the tests results should be deterministic (excluding collectives like reduction on devices), correct?

forrestglines commented:
> Just do double check: the tests results should be deterministic (excluding collectives like reduction on devices), correct?

Yes, they should be deterministic. I haven't included any random perturbations in these. If they aren't then there's a problem.

forrestglines commented:
> Are they failing consistently or at different times/blocks?

I will need to check this later. As I recall, no

pgrete commented on src/hydro/hydro_driver.cpp:
```diff
@@ -86,6 +88,45 @@ TaskCollection HydroDriver::MakeTaskCollection(BlockList_t &blocks, int stage) {
+  const int num_partitions = pmesh->DefaultNumPartitions();
+
+  // calculate agn triggering accretion rate
+  if ((stage == 1) &&
+      hydro_pkg->AllParams().hasKey("agn_triggering_reduce_accretion_rate") &&
+      hydro_pkg->Param<bool>("agn_triggering_reduce_accretion_rate")) {
+
+    // need to make sure that there's only one region in order to MPI_reduce to work
+    TaskRegion &single_task_region = tc.AddRegion(1);
+    auto &tl = single_task_region[0];
+    // First globally reset triggering quantities
+    auto prev_task =
+        tl.AddTask(none, cluster::AGNTriggeringResetTriggering, hydro_pkg.get());
+
+    // Adding one task for each partition. Given that they're all in one task list
+    // they'll be executed sequentially. Given that a par_reduce to a host var is
+    // blocking it's also save to store the variable in the Params for now.
+    for (int i = 0; i < num_partitions; i++) {
+      auto &mu0 = pmesh->mesh_data.GetOrAdd("base", i);
+      auto new_agn_triggering =
+          tl.AddTask(prev_task, cluster::AGNTriggeringReduceTriggering, mu0.get(), tm.dt);
+      prev_task = new_agn_triggering;
+    }
+#ifdef MPI_PARALLEL
+    auto reduce_agn_triggering =
+        tl.AddTask(prev_task, cluster::AGNTriggeringMPIReduceTriggering, hydro_pkg.get());
+    prev_task = reduce_agn_triggering;
+#endif
```
We can leave this for now but should open an issue to eval if this could be replaced with a (clean) Parthenon MPI parallel reduction

forrestglines replied:
I agree. I wrote this code before that reduction existed in Parthenon.

forrestglines replied:
Made #62 to address this later

pgrete commented on src/hydro/hydro_driver.cpp:
```diff
@@ -86,6 +88,45 @@ TaskCollection HydroDriver::MakeTaskCollection(BlockList_t &blocks, int stage) {
+    }
+  }
+
+  TaskRegion &async_region_1 = tc.AddRegion(num_task_lists_executed_independently);
```
Any reason for this new region?

forrestglines replied:
I'm guessing this is left over changes from merging in `main` several times.

pgrete commented on src/hydro/hydro_driver.cpp:
```diff
@@ -150,6 +189,48 @@ TaskCollection HydroDriver::MakeTaskCollection(BlockList_t &blocks, int stage) {
 
+  // calculate magnetic tower scaling
+  if ((stage == 1) && hydro_pkg->AllParams().hasKey("magnetic_tower_power_scaling") &&
+      hydro_pkg->Param<bool>("magnetic_tower_power_scaling")) {
+    const auto &magnetic_tower =
+        hydro_pkg->Param<cluster::MagneticTower>("magnetic_tower");
+
+    // need to make sure that there's only one region in order to MPI_reduce to work
+    TaskRegion &single_task_region = tc.AddRegion(1);
+    auto &tl = single_task_region[0];
+    // First globally reset magnetic_tower_linear_contrib and
+    // magnetic_tower_quadratic_contrib
+    auto prev_task =
+        tl.AddTask(none, cluster::MagneticTowerResetPowerContribs, hydro_pkg.get());
+
+    // Adding one task for each partition. Given that they're all in one task list
+    // they'll be executed sequentially. Given that a par_reduce to a host var is
+    // blocking it's also save to store the variable in the Params for now.
+    for (int i = 0; i < num_partitions; i++) {
+      auto &mu0 = pmesh->mesh_data.GetOrAdd("base", i);
+      auto new_magnetic_tower_power_contrib =
+          tl.AddTask(prev_task, cluster::MagneticTowerReducePowerContribs, mu0.get(), tm);
+      prev_task = new_magnetic_tower_power_contrib;
+    }
+#ifdef MPI_PARALLEL
+    auto reduce_magnetic_tower_power_contrib = tl.AddTask(
+        prev_task,
+        [](StateDescriptor *hydro_pkg) {
+          Real magnetic_tower_contribs[] = {
+              hydro_pkg->Param<Real>("magnetic_tower_linear_contrib"),
+              hydro_pkg->Param<Real>("magnetic_tower_quadratic_contrib")};
+          PARTHENON_MPI_CHECK(MPI_Allreduce(MPI_IN_PLACE, magnetic_tower_contribs, 2,
+                                            MPI_PARTHENON_REAL, MPI_MAX, MPI_COMM_WORLD));
+          hydro_pkg->UpdateParam("magnetic_tower_linear_contrib",
+                                 magnetic_tower_contribs[0]);
+          hydro_pkg->UpdateParam("magnetic_tower_quadratic_contrib",
+                                 magnetic_tower_contribs[1]);
+          return TaskStatus::complete;
+        },
+        hydro_pkg.get());
```
As above, if this keeps popping up as a pattern we should abstract it (but keep it for now)

pgrete commented on src/hydro/hydro_driver.cpp:
```diff
@@ -150,6 +189,48 @@ TaskCollection HydroDriver::MakeTaskCollection(BlockList_t &blocks, int stage) {
+              hydro_pkg->Param<Real>("magnetic_tower_linear_contrib"),
+              hydro_pkg->Param<Real>("magnetic_tower_quadratic_contrib")};
+          PARTHENON_MPI_CHECK(MPI_Allreduce(MPI_IN_PLACE, magnetic_tower_contribs, 2,
+                                            MPI_PARTHENON_REAL, MPI_MAX, MPI_COMM_WORLD));
```
Potential bug? `MagneticTowerReducePowerContribs` uses uses a sum reducer, but this is `MPI_MAX`

forrestglines replied:
Definitely. I didn't catch this since I didn't add an MPI version of the triggering test.

forrestglines replied:
I tried adding a `setup_test_parallel` version of these tests but ran into issues with `MPI_FOUND` not escaping the scope of the `external/parthenon/CMakeLists.txt`. I didn't know how to fix that.

I did test the `cluster_magnetic_tower` test with MPI before and after the change, which confirmed there was an MPI bug and this fixes it.

pgrete reviewed (commented):
First batch of comments. Will continue over the next days, but feel free to already reply.
Also, I'll answer the questions you pose in the PR description when I'm done reviewing because I'll have a better sense of the code then.

pgrete commented:
Also general comment: Do you plan to add any doc to the `doc` folder?

forrestglines commented on inputs/cluster/agn_triggering.in:
```diff
@@ -0,0 +1,114 @@
+
+<problem/cluster/gravity>
+#Define SMBH for Bondi accretion
+m_smbh = 1.0000000000000002e-06
```
The mass is 1e-6 code-mass-units. 

Having mixed units in parameter files - some things in kpc, other in cgs, others in msun, etc. turned out to be a horrible mess in Enzo that lead to several user-errors on my part and previous developers. There were code-unit conversions everywhere and it made it difficult to debug/track parameters defined in the input file and where they were used in the code. With this implementation everything is in code-units.

I'm in favor of having everything in one system of units in parameter files. Thus, everything in the cluster input file is in code units except for the actual definition of code units.

I use a Jupyter notebook to generate and manage the input parameters for clusters, thus the floating point round-off. This has the added benefit to estimate data and memory usage when I create the file as well as potentially making plots of the initial cluster.

pgrete replied:
Understood. As long as we're consistent and explicit I have not strong opinion on which way to go.
Especially for these more complex scenarios an external path (like the jupyter notebooks you prepared) may be more user friendly (also with regard to further analysis) than handling all the logic internally.
Having said that: How about to add the setup notebook also to the repo so that there's only one place where this lives (and needs to be maintained -- similar to the `generate forcing wavenumbers` python script I use for the turbulence sims).

forrestglines commented on inputs/cluster/hse.in:
```diff
@@ -73,47 +73,54 @@ He_mass_fraction = 0.25
 #NFW parameters
 c_nfw = 6.0
-M_nfw_200 = 10.000000000000002
+m_nfw_200 = 10.000000000000002
```
I generated these with a python script which introduced this round off. We can chop that off for these testing files.

forrestglines commented on inputs/cluster/agn_triggering.in:
```diff
@@ -0,0 +1,114 @@
+<problem/cluster/uniform_gas>
+#Initialize with a uniform gas
+init_uniform_gas = true
+rho = 14775.575892787232
+ux = 0.0006136272991326239
+uy = 0.0004090848660884159
+uz =-0.0005113560826105199
+pres = 1.5454368403867562
```
I prefer using code-units as much as possible so that it's unambiguous what units are being used and so that for debugging purposes the value in the code should match the value in the input file.

forrestglines commented on inputs/cluster/sutherland_dopita.cooling:
```diff
@@ -0,0 +1,782 @@
+# Cooling table for solar metallicity, 1/3 solar metallicity
```
I think that's a great place, I'll implement that.

forrestglines commented on inputs/cluster/cluster.in:
```diff
@@ -0,0 +1,192 @@
+<problem/cluster/agn_triggering>
+triggering_mode = COLD_GAS
+#NOTE: Change to this line to disable AGN triggering
+#triggering_mode = NONE
```
I slightly prefer all-caps to match conventions for all cap enums but I'd be fine with other caps. DISABLED would be more descriptive here. Should I change that?

pgrete replied:
I'm in slight favor of lowercase mostly because the majority of parameters in the input file (also for parthenon itself) are lowercase.

forrestglines replied:
Lets punt this to a parameter refactor in #24

forrestglines commented on src/eos/eos.hpp:
```diff
@@ -46,7 +46,7 @@ class EquationOfState {
   Real GetInternalEFloor() const { return internal_e_floor_; }
 
- private:
+ protected:
```
It was necessary to access those member floors within `ConsToPrim` kernels. However, I see that these are exposed via inline functions and so those member functions can just be used within the equations of state that inherit `EquationOfState`

forrestglines commented on src/eos/adiabatic_glmmhd.hpp:
```diff
@@ -52,6 +52,98 @@ class AdiabaticGLMMHDEOS : public EquationOfState {
+  // \!fn Real EquationOfState::ConsToPrim(View4D cons, View4D prim, const int& k, const
+  // int& j, const int& i) \brief Fills an array of primitives given an array of
+  // conserveds, potentially updating the conserved with floors
+  template <typename View4D>
+  KOKKOS_INLINE_FUNCTION void ConsToPrim(View4D cons, View4D prim, const int &k,
```
It should be, yes. I'll change it.

forrestglines commented on src/eos/adiabatic_glmmhd.hpp:
```diff
@@ -52,6 +52,98 @@ class AdiabaticGLMMHDEOS : public EquationOfState {
+    w_p = gm1 * (u_e - e_k - e_B);
+
+    // apply pressure floor, correct total energy
+    u_e = (w_p > pressure_floor_) ? u_e : ((pressure_floor_ / gm1) + e_k + e_B);
+    w_p = (w_p > pressure_floor_) ? w_p : pressure_floor_;
```
I'll consolidate the functions into the inlined function in the header file.

Can we apply both floors, one after the other, so that whichever floor is more constrictive is applied?

forrestglines replied:
While doing this, I've realized passive scalars have been added since I wrote this so more changes will be required to correctly consolidate things.

pgrete replied:
> Can we apply both floors, one after the other, so that whichever floor is more constrictive is applied?

Definitely. We should discuss how to approach this in general (and how much runtime flexibility we want)

forrestglines commented on src/eos/adiabatic_glmmhd.hpp:
```diff
@@ -52,6 +52,98 @@ class AdiabaticGLMMHDEOS : public EquationOfState {
+  // int& j, const int& i) \brief Fills an array of conserveds given an array of
+  // primitives,
+  template <typename View4D>
+  KOKKOS_INLINE_FUNCTION void PrimToCons(View4D prim, View4D cons, const int &k,
+                                         const int &j, const int &i) const {
```
Before I forget this issue, these functions do not enforce floors. Should this function also modify `prim` to enforce floors like `ConsToPrim` can modify `cons`? Probably?

forrestglines replied:
Actually, I ended up not using these functions at all. I will remove them for now

forrestglines commented:
@par-hermes format

forrestglines commented on inputs/cluster/hydro_agn_feedback.in:
```diff
@@ -0,0 +1,114 @@
+gamma = 1.6666666666666667 # gamma = C_p/C_v
+eos = adiabatic
+riemann = none
+reconstruction = dc
```
We're not evolving the fluid in these tests, I don't think reconstruction matters here. I get this error message when I try:
```
b"### PARTHENON ERROR
  Condition:   recon == Reconstruction::dc
  Message:     Disabling hyperbolic fluxes via 'none' Riemann solver only supported in comination with DC reconstruction.
  File:        /vast/home/forrestglines/code/athenapk-project/athenapk/src/hydro/hydro.cpp
  Line number: 282
"
```

pgrete replied:
Oh right, I missed the `riemann=none` part.
Might be worth to expand a little at the top of the file what this test case does as `hydro_agn_feedback` makes my brain expect that sth is going to evolve.

forrestglines commented on inputs/cluster/agn_triggering.in:
```diff
@@ -0,0 +1,114 @@
+nlim       = -1         # cycle limit
+tlim       = 0.1       # time limit
+integrator  = vl2       # time integration algorithm
+perf_cycle_offset  = 10 # interval for stdout summary info
```
I copied this from somewhere but it's not necessary for the test. I'll just remove this parameter

forrestglines commented:
The test failures on the A40s on Delta GPU appear to be related to using static refinement and GLMMHD, with smaller mesh blocks leading to a higher frequency of errors. I've isolated the failure in a linear wave test using the `main` branch and created an issue in #10

forrestglines commented on inputs/cluster/hydro_agn_feedback.in:
```diff
@@ -0,0 +1,114 @@
+
+<parthenon/output2>
+file_type  = hdf5      # HDF5 data dump
+variables  = cons,prim      # Variables to be output
```
To have both sets of variables available for testing without needing to implement primitive to conserved conversions in testing.

forrestglines commented:
> Also general comment: Do you plan to add any doc to the `doc` folder?

Yes, I need to do this before this is truly ready to be merged

forrestglines commented:
One programmatic/logical bug I've realized is that due to the way calculating AGN triggering via hijacking the history reduction in Parthenon, the AGN power is set to zero for the first timestep running with an input file but more concerningly after restarts.

This is due to history reductions happening at the end of a timestep when the AGN triggering should be computed at the beginning of the timestep. It might be better to expedite changing the AGN triggering to use the newer Parthenon reductions.

pgrete commented:
@forrestglines I just updated to current `main`, i.e., with the new coordiantes and updated Parthenon/Kokkos submodule.
Would you mind having a quick look at the commit https://github.com/parthenon-hpc-lab/athenapk/pull/2/commits/a671f92c1da36c348d35c37598897b537a7d53b1 to see if I have missed sth obvious (as I directly pushed here)?

forrestglines commented:
> @forrestglines I just updated to current `main`, i.e., with the new coordiantes and updated Parthenon/Kokkos submodule. Would you mind having a quick look at the commit [a671f92](https://github.com/parthenon-hpc-lab/athenapk/commit/a671f92c1da36c348d35c37598897b537a7d53b1) to see if I have missed sth obvious (as I directly pushed here)?

Looks good, it's running fine.

pgrete commented on src/eos/adiabatic_glmmhd.cpp:
```diff
@@ -37,88 +37,19 @@ void AdiabaticGLMMHDEOS::ConservedToPrimitive(MeshData<Real> *md) const {
   const auto nhydro = pkg->Param<int>("nhydro");
   const auto nscalars = pkg->Param<int>("nscalars");
 
+  auto this_on_device = (*this);
```
This is required to get access to the `GetXXX` calls, doesn't it?
Have you profiled how much overhead is introduced by those calls?
Previously, these were captured scalars. Now these functions are called for each index.

pgrete commented on src/hydro/srcterms/tabular_cooling.cpp:
```diff
@@ -392,11 +393,18 @@ void TabularCooling::SubcyclingFixedIntSrcTerm(MeshData<Real> *md, const Real dt
+                //PARTHENON_FAIL("FATAL ERROR in [TabularCooling::SubcyclingSplitSrcTerm]: "
+                //               "Minumum sub_dt leads to negative internal energy");
+                
+                //HACK
+                //Even the minimum subcycle dt would lead to negative internal energy -- so just cool to the floor
```
Is this physically correct, i.e., is the case where this is triggered always meaning that the cooling should go down to the floor or could it also be that this may happen higher up in the cooling table where there's just a bad subcycle (but the final temperature would not be at the floor)?
Also, if we've properly limited `dt` (say the default 10% fraction of change in internal energy) this should not be triggered or am I missing sth?

forrestglines replied:
This section of code will be triggered whenever a negative energy or NAN energy is encountered within any of the RK stages in the current subcycle and only when using the minimum subcycle timestep allowed.  I believe as long as there's a minimum timestep restriction on subcycles, this is the "last change" subcycle so not just any bad subcycle. 

This would lead to an incorrect final temperature if you had a cooling with a high cooling rate a few orders of magnitude above the temperature floor but a long asymptote to zero going down the the temperature flow. In practice though with the cooling tables I've been using, the cooling rate above the temperature floor is fast enough to drop to zero during this minimum subcycle timestep.

My best idea here is to check whether a subcycle at the current subcycle internal energy and minimum subcycle timestep would lead to a negative energy. If so, just cool to the floor. It might not be perfect, but I also don't have much interest trying to support  a contrived cooling table made to break the inferior cooling integrator when one can just use the `Townsend` integrator. 

You are also correct that this code is never encountered when `EstimateTimeStep` is enabled via `cooling/cfl=0.1` as you describe. During testing, however, I found that limiting the hydro timestep by the minimum cooling time led to the simulation to come to a halt when cold gas was forming. When I disabled the "cooling cfl", it lead to the commented out exception and my hack to get around it.

forrestglines replied:
This has been updated/addressed in #28, we'll need to merge those changes into this branch.

pgrete commented on src/hydro/srcterms/tabular_cooling.cpp:
```diff
@@ -460,7 +468,11 @@ void TabularCooling::SubcyclingFixedIntSrcTerm(MeshData<Real> *md, const Real dt
 
-        PARTHENON_REQUIRE(internal_e > internal_e_floor, "cooled below floor");
+        //PARTHENON_REQUIRE(internal_e > internal_e_floor, "cooled below floor");
+        //HACK
+        if(internal_e < internal_e_floor){
+          internal_e = internal_e_floor;
+        }
```
Should this remain? Or in other words, should this be safety check (for this subcycling) or an early "fix" what would otherwise later be triggered in `ConsToPrim`?

forrestglines replied:
I saw issues for negative energies before this line was added. If that fix isn't already in `ConsToPrim` but just wasn't working, we could put it there as long as other source terms after this don't depend on `Cons` or `Prim`

forrestglines replied:
This should be addressed correctly in #28

pgrete commented on src/hydro/srcterms/tabular_cooling.hpp:
```diff
@@ -116,6 +116,9 @@ class TabularCooling {
   CoolIntegrator integrator_;
 
   // Temperature floor (assumed in Kelvin and only used in cooling function)
+  // This is either the temperature floor used by the hydro method or the
+  // lowest temperature in the cooling table (assuming zero cooling below the
+  // table), whichever temperature is higher
```
We might want to discuss if it'd make sense to introduce a distinction between those "floors".
For example, we may want to have expanding plasma that "cools" (so actually can get colder without active cooling) without being reset to the floor.

forrestglines replied:
This has been fixed in #28, correct?

pgrete commented on src/hydro/srcterms/tabular_cooling.hpp:
```diff
@@ -150,14 +157,16 @@ class TabularCooling {
+      // TODO(forrestglines):Currently, zero cooling is use for temperatures
+      // below the table. This behavior could be generalized via templates
+      // return 0;
+      log_lambda = log_lambdas(0); // HACK - to get it to cool to floor
```
Related to the comment above. We should discuss next week during the call what we physically want (in terms of floors, choice of cooling table, etc).

pgrete commented on src/pgen/cluster.cpp:
```diff
@@ -166,15 +227,13 @@ void ProblemGenerator(MeshBlock *pmb, parthenon::ParameterInput *pin) {
-        for (int i = ib.s; i <= ib.e; i++) {
-
+    parthenon::par_for(
+        DEFAULT_LOOP_PATTERN, "cluster::ProblemGenerator::UniformGas",
```
This is the else branch of the `UniformGas` conditional, isn't it?

forrestglines replied:
Yes, I can add a comment to clarify that.

pgrete reviewed (commented):
Another "first" bunch of comments. Half way through the files now.

forrestglines commented on .gitignore:
```diff
@@ -37,3 +37,7 @@ compile_commands.json
 Testing
+
+# generate_cluster_input.ipynb files
+inputs/cluster/generate_cluster_input.ipynb
```
I put this file into `.gitignore` not to exclude any future changes to the notebook, but to exclude it from a `git add -A` and `git status`. Running the notebook will change the notebook, rerunning cells in the notebook in different orders will also change the notebook, and I don't want those changes added automatically or negligently (most likely by me). I also don't want other users accidentally uploading their changes to input files as they experiment. You can only make changes to this file deliberately with `git add --force`

forrestglines commented on inputs/cluster/generate_cluster_input.ipynb:
```diff
@@ -0,0 +1,569 @@
+    "    \"\"\"\n",
+    "    Helper function to quickly define static-mesh refinement meshes for AthenaPK.\n",
+    "    By default, prints out information like smallest cell size, total number of\n",
+    "    cells, estimated data outputs, and estimated NVIDIA A100s needed to run the\n",
```
Agreed. I nabbed these numbers from my own empirical testing of the A100s on Delta. We can be more precise

forrestglines commented on docs/cluster.md:
````diff
@@ -0,0 +1,396 @@
+This profile is determined by these parameters
+```
+<problem/cluster/entropy_profile>
+k_0 = 8.851337676479303e-121 # in FIXME
+k_100 = 1.3277006514718954e-119 # in FIXME
````
I need to add the units of these in code units for the FIXME. You're correct that we need `Real` to be `float64` here

forrestglines replied:
I'm not sure how to bring this quantity closer to unity, perhaps divided by mean particle mass. This could be explored in a cleanup PR later.

For now, the correct units have been added to the documentation

forrestglines commented on docs/cluster.md:
````diff
@@ -0,0 +1,396 @@
+accretion_radius = 0.001 # in code_length
+bondi_alpha= 100.0 # unitless
+```
+With BONDI_SCHAYE accretion, the `alpha` used for BOOSTED_BONDI accretion is modified to depend on the number density following:
````
"$\alpha$", I'll fix that

pgrete commented on src/pgen/cluster/agn_triggering.cpp:
```diff
@@ -0,0 +1,626 @@
+    triggering_file.close();
+  }
+
+  // Set up UserHistoryOutput (Are all these triggering variables necessary?
```
I think we agreed that it's not necessary.

pgrete commented on src/pgen/cluster/agn_triggering.cpp:
```diff
@@ -0,0 +1,626 @@
+            const Real cell_cold_mass = prim(IDN, k, j, i) * coords.CellVolume(k, j, i);
+            team_cold_mass += cell_cold_mass;
+
+            const Real cell_delta_rho = -prim(IDN, k, j, i) / cold_t_acc * dt;
```
We should double check what `dt` is here. Depending on how it's called we might need to use `beta_dt`

pgrete replied:
Yes, correct.

pgrete commented on src/pgen/cluster/agn_triggering.cpp:
```diff
@@ -0,0 +1,626 @@
+            if (remove_accreted_mass) {
+              AddDensityToConsAtFixedVelTemp(cell_delta_rho, cons, prim, eos, k, j, i);
+              // Update the Primitives
+              eos.ConsToPrim(cons, prim, nhydro, nscalars, k, j, i);
```
Reminder to myself: double check this is the intended behavior (in terms of where this function is called in the task list)

pgrete replied:
POTENTIAL BUG (should discuss today)
If we update the variables here, we should also do it in the ghost zones as the `interior` info is currently not communicated any more.

forrestglines replied:
I've updated both the cold gas triggering and Bondi accretion to remove mass from the entire grid, not just interior grid. The reductions are still done over the interior grid so as to not double count gas.

pgrete commented on src/pgen/cluster/agn_triggering.cpp:
```diff
@@ -0,0 +1,626 @@
+    // Test the Bondi-like triggering methods
+    const Real total_mass = hydro_pkg->Param<Real>("agn_triggering_total_mass");
+    if (total_mass == 0) {
+      // TODO(forrestglines)During the first timestep, the total mass and
+      // accretion rate has not yet been reduced. However, since accreted mass is
+      // removed during that reduction, the timestep is needed to execute that
+      // reduction. As a compromise, we ignore the timestep constraint during the
```
Isn't this even more general than just the first timestep?
I.e., if `total_mass==0` nothing should happen independent where in the simulation we are, doesn't it?

forrestglines replied:
If `total_mass==0` after the first reduction, then the total mass in the accretion zone is 0 and something must be broken. Only between its initialization and the first reduction should `total_mass==0`. Nonzero `total_mass` should carry over with restarts.

pgrete commented on src/pgen/cluster/cluster_utils.hpp:
```diff
@@ -0,0 +1,88 @@
+
+namespace cluster {
+
+// Add a density to the conserved variables while keeping velocity fixed
+template <typename View4D>
+KOKKOS_INLINE_FUNCTION void
+AddDensityToConsAtFixedVel(const parthenon::Real density, View4D &cons,
+                           const View4D &prim, const AdiabaticHydroEOS &eos, const int &k,
+                           const int &j, const int &i) {
+  // Add density such that velocity and temperature (propto pressure/density) is fixed
+  cons(IDN, k, j, i) += density;
+  cons(IM1, k, j, i) += density * prim(IV1, k, j, i);
+  cons(IM2, k, j, i) += density * prim(IV2, k, j, i);
+  cons(IM3, k, j, i) += density * prim(IV3, k, j, i);
+  cons(IEN, k, j, i) +=
+      density * (0.5 * (SQR(prim(IV1, k, j, i)) + SQR(prim(IV2, k, j, i)) +
+                        SQR(prim(IV3, k, j, i))));
+}
+
+template <typename View4D>
+KOKKOS_INLINE_FUNCTION void
+AddDensityToConsAtFixedVel(const parthenon::Real density, View4D &cons,
+                           const View4D &prim, const AdiabaticGLMMHDEOS &eos,
+                           const int &k, const int &j, const int &i) {
+  // Add density such that velocity and temperature (propto pressure/density) is fixed
+  cons(IDN, k, j, i) += density;
+  cons(IM1, k, j, i) += density * prim(IV1, k, j, i);
+  cons(IM2, k, j, i) += density * prim(IV2, k, j, i);
+  cons(IM3, k, j, i) += density * prim(IV3, k, j, i);
+  cons(IEN, k, j, i) +=
+      density * (0.5 * (SQR(prim(IV1, k, j, i)) + SQR(prim(IV2, k, j, i)) +
+                        SQR(prim(IV3, k, j, i))));
+}
```
No need to differentiate between those two calls.

forrestglines replied:
Changed to one function here, the `AddDensityToConsAtFixedVelTemp` to take an adiabatic index instead of an EOS.

pgrete commented on src/pgen/cluster/cluster_utils.hpp:
```diff
@@ -0,0 +1,88 @@
+  cons(IM1, k, j, i) += density * prim(IV1, k, j, i);
+  cons(IM2, k, j, i) += density * prim(IV2, k, j, i);
+  cons(IM3, k, j, i) += density * prim(IV3, k, j, i);
+  cons(IEN, k, j, i) +=
+      density * (0.5 * (SQR(prim(IV1, k, j, i)) + SQR(prim(IV2, k, j, i)) +
+                        SQR(prim(IV3, k, j, i))) +
+                 1 / (eos.GetGamma() - 1.0) * prim(IPR, k, j, i) / prim(IDN, k, j, i));
```
Reminder to myself: double check this yields desired behavior.

pgrete reviewed (commented):
More comments. Not done yet.

pgrete reviewed (commented):
Alright, I'm done with the "code" review, i.e, I now have looked at the code itself.
Before merging I'd like to run a couple of more local tests to also "review" the integrated physics. 
I plan to do so next week. Feel free to (continue to) reply to as many comments as feasible.

In addition to the ones I left attached to individual lines, here are some more general ones:
- `use_scratch = false` does not exist any more
- Follow up PR: cleanup tasks (by moving all "pgen" related calls to callback functions, which may need to be introduced)
- The setup should just be run with a single pack for now (I need to double check the logic for `GetOrAdd`)
- Discuss what to do with `Tfloor` and cooling (lower end of the table)
- General review of open TODOs in codebase
- Follow up PR: Cleanup package/parameter/global variable handling (may require upstream changes), e.g., call to `agn_triggering.GetAccretionRate(hydro_pkg);` where `agn_triggering` was obtained from `hydro_pkg` two lines above.
- Follow up PR: Cleanup/centralize plasma composition related code, e.g. calculating `mu` or `
ne_from_rho`
- Discuss if we want to properly write triggering log every cycle (i.e., MPI support). Should be separate PR.

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,230 @@
+
+parthenon::Real AGNFeedback::GetFeedbackPower(StateDescriptor *hydro_pkg) const {
+  auto units = hydro_pkg->Param<Units>("units");
+  const auto &agn_triggering = hydro_pkg->Param<AGNTriggering>("agn_triggering");
+
+  const Real accretion_rate = agn_triggering.GetAccretionRate(hydro_pkg);
```
I'd be open to cleaning that up. 

A pattern I've been using is to make a different object that keeps track of its internal state for each module of the cluster setup. I think `hydro_pkg` is available at the creation of these objects, so they could store a pointer to `hydro_pkg`. That would mean an object with a pointer to `hydro_pkg` contained within a map in `hydro_pkg`, but I don't think that's actually any issue.

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,230 @@
+                                  const parthenon::SimTime &tm) const {
+  auto hydro_pkg = md->GetBlockData(0)->GetBlockPointer()->packages.Get("Hydro");
+  auto fluid = hydro_pkg->Param<Fluid>("fluid");
+  if (fluid == Fluid::euler) {
+    FeedbackSrcTerm(md, beta_dt, tm, hydro_pkg->Param<AdiabaticHydroEOS>("eos"));
+  } else if (fluid == Fluid::glmmhd) {
+    FeedbackSrcTerm(md, beta_dt, tm, hydro_pkg->Param<AdiabaticGLMMHDEOS>("eos"));
+  } else {
+    PARTHENON_FAIL("AGNFeedback::FeedbackSrcTerm: Unknown EOS");
+  }
```
I'd be fine with that, although I prefer the template to be right here. That way all source terms retain the external syntax `SrcTerm(MeshData,dt,SimTime)` and they're all called the same way in `cluster.cpp`

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,230 @@
+  // Grab some necessary variables
+  const auto &prim_pack = md->PackVariables(std::vector<std::string>{"prim"});
+  const auto &cons_pack = md->PackVariables(std::vector<std::string>{"cons"});
+  IndexRange ib = md->GetBlockData(0)->GetBoundsI(IndexDomain::interior);
```
I agree, that would be much cleaner and I've been wanting something like that.

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,230 @@
+
+  const auto &jet_coords_factory =
+      hydro_pkg->Param<JetCoordsFactory>("jet_coords_factory");
+  const JetCoords jet_coords = jet_coords_factory.CreateJetCoords(time);
```
Yes, it's a very lightweight object. It only contains the current direction of the jet.

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,230 @@
+          // Determine if point is in sphere r<=thermal_radius
+          if (r2 <= thermal_radius2) {
+            // Add density at constant velocity and temperature
+            // HACK Disabling injection of mass with thermal feedback since it
+            // leads to small timesteps
+            //AddDensityToConsAtFixedVelTemp(thermal_density_rate, cons, prim, eos, k, j,i);
```
Probably? There's a question about strict accounting for mass removed during AGN triggering. Should we inject that mass back in with feedback minus the mass converted to energy? Maybe take some of that mass and add it to the SMBH mass?

For the commit here, I've been adding that mass back proportionally to the feedback mechanisms (i.e., 1/3 the energy goes into kinetic energy, 1/3 the mass gets injected with the kinetic jet). I hacked off the thermal mass injection because there was a bug, the thermal density injection was missing `dt`. 

I've fixed that bug in a commit I was making and testing last week. I also modified the mass injection to account for the mass-to-energy conversion (which is negligible) but more importantly for the proportion of mass injection to be different  from energy injection (i.e. 1/3 energy into the kinetic jet, 1/8 the mass into the kinetic jet). This turns out to be necessary to change the jet velocity due by disconnecting the kinetic energy injected and mass injected.

On the topic of injecting mass with thermal feedback, it might avoid some low density scenarios. Dumping thermal energy near the AGN it tends to blow out the gas. Maybe by injecting some of the AGN triggering gas we can avoid that.

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,230 @@
+
+            const int sign_jet = (h > 0) ? 1 : -1; // Above or below jet-disk
+
+            cons(IDN, k, j, i) += kinetic_jet_density_rate; // mass/volume
+            cons(IM1, k, j, i) += kinetic_jet_momentum_rate * sign_jet * jet_axis_x;
+            // velocity*mass/volume
+            cons(IM2, k, j, i) += kinetic_jet_momentum_rate * sign_jet * jet_axis_y;
+            // velocity*mass/volume
+            cons(IM3, k, j, i) += kinetic_jet_momentum_rate * sign_jet * jet_axis_z;
+            // velocity*mass/volume
+            cons(IEN, k, j, i) += kinetic_feedback; // energy/volume
```
I'm not sure it is. Currently, in each cell, we're adding a new mass with density in this cell `kinetic_jet_density`. We add that mass with velocity `kinetic_jet_velocity` along the vector `sign_jet *jet_axis`. Thus the change in momentum density is `kinetic_density*kinetic_jet_velocity*sign_jet *jet_axis`. We then add the kinetic energy per volume `kinetic_jet_density*kinetic_jet_velocity**2`. 

However, when you compare the kinetic energy added versus the change in `1./2*momentum**2/mass`, they don't match. I think we need to solve for the change in momentum from the mass added and kinetic energy added.

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,230 @@
+
+  // Add user history output variable for AGN power
+  auto hst_vars = hydro_pkg->Param<parthenon::HstVar_list>(parthenon::hist_param_key);
+  if (!disabled_) {
+    // HACK (forrestglines): The operations should be a
+    // parthenon::UserHistoryOperation::no_reduce, which is as of writing
+    // unimplemented
+    hst_vars.emplace_back(parthenon::HistoryOutputVar(
+        parthenon::UserHistoryOperation::max,
+        [this](MeshData<Real> *md) {
+          auto pmb = md->GetBlockData(0)->GetBlockPointer();
+          auto hydro_pkg = pmb->packages.Get("Hydro");
+          return this->GetFeedbackPower(hydro_pkg.get());
+        },
+        "agn_feedback_power"));
+  }
+  hydro_pkg->UpdateParam(parthenon::hist_param_key, hst_vars);
```
Yes, we do want the AGN feedback power in the history output. It's not necessary for simulation correctness but it would be essentially necessary for analysis.

There was discussion of implementing a `no_reduce` history operation that wouldn't do any reduction/barrier for this variable.

I think that said performance fix, especially for the other variables, would be great. Theoretically these could all be nonblocking reductions with one barrier after all the history reductions?

forrestglines commented on src/pgen/cluster/hydrostatic_equilibrium_sphere.hpp:
```diff
@@ -56,35 +58,38 @@ class HydrostaticEquilibriumSphere {
+                                                      const parthenon::Real k) const {
     const parthenon::Real rho =
-        pow(P / K, 3. / 5.) * mu_ * atomic_mass_unit_ / pow(mu_ / mu_e_, 2. / 5);
+        pow(p / k, 3. / 5.) * mu_ * atomic_mass_unit_ / pow(mu_ / mu_e_, 2. / 5);
```
Yes, I had to double check that this is dimensional. I've fixed up the math a little be to remove redundant terms.

forrestglines commented on src/pgen/cluster/hydrostatic_equilibrium_sphere.hpp:
```diff
@@ -56,35 +58,38 @@ class HydrostaticEquilibriumSphere {
 
   // Get total number density from density
-  parthenon::Real n_from_rho(const parthenon::Real rho) const {
+  KOKKOS_INLINE_FUNCTION parthenon::Real n_from_rho(const parthenon::Real rho) const {
     const parthenon::Real n = rho / (mu_ * atomic_mass_unit_);
     return n;
   }
 
   // Get electron number density from density
-  parthenon::Real ne_from_rho(const parthenon::Real rho) const {
+  KOKKOS_INLINE_FUNCTION parthenon::Real ne_from_rho(const parthenon::Real rho) const {
     const parthenon::Real ne = mu_ / mu_e_ * n_from_rho(rho);
     return ne;
   }
```
Agreed. Many of these terms are general for a plasma with cosmic abundances of hydrogen and helium. Most are only used to generate hydrodstatic equilibrium from the "entropy" K profile.

forrestglines commented on src/pgen/cluster/magnetic_tower.hpp:
```diff
@@ -0,0 +1,193 @@
+                                  const View4D &A) const;
+
+  // Add the fixed_field_rate  (and associated magnetic energy) to the
+  // conserved variables for all meshblocks with a MeshData
```
That's a typo, I mean "within" not "with" here.

forrestglines commented on src/reduction_utils.hpp:
```diff
@@ -0,0 +1,65 @@
+
+#ifndef REDUCTION_UTILS_HPP_
+#define REDUCTION_UTILS_HPP_
+// Consider moving ReductionSumArray to Parthenon and extending to other operations
```
That construct is new, right? That sounds like it does exactly what we want.

forrestglines replied:
I can't get that functionality to compile. My code in `agn_triggering.cpp` looks like this. If we can get it to compile we can remove this file
```c++
   // Reduce Mass-weighted total density, velocity, and sound speed and total
  // mass (in that order). Will need to divide the three latter quantities by
  // total mass in order to get their mass-weighted averaged values

  const parthenon::Real gamma = gamma_;

  parthenon::par_reduce(
      parthenon::loop_pattern_mdrange_tag, "AGNTriggering::ReduceBondi",
      parthenon::DevExecSpace(), 0, cons_pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s,
      ib.e,
      KOKKOS_LAMBDA(const int &b, const int &k, const int &j, const int &i, 
        Real &team_total_mass, Real &team_mw_density, Real &team_mw_velocity, Real &team_mw_cs ) { 

        auto &cons = cons_pack(b);
        auto &prim = prim_pack(b);
        const auto &coords = cons_pack.GetCoords(b);
        const parthenon::Real r2 =
            pow(coords.Xc<1>(i), 2) + pow(coords.Xc<2>(j), 2) + pow(coords.Xc<3>(k), 2); 
        if (r2 < accretion_radius2) {
          const Real cell_mass = prim(IDN, k, j, i) * coords.CellVolume(k, j, i); 

          const Real cell_mw_density = cell_mass * prim(IDN, k, j, i); 
          const Real cell_mw_velocity =
              cell_mass * sqrt(pow(prim(IV1, k, j, i), 2) + pow(prim(IV2, k, j, i), 2) +
                               pow(prim(IV3, k, j, i), 2));
          const Real cell_mw_cs =
              cell_mass * sqrt(gamma * prim(IPR, k, j, i) / prim(IDN, k, j, i));

          team_total_mass += cell_mass;
          team_mw_density += cell_mw_density;
          team_mw_velocity += cell_mw_velocity;
          team_mw_cs += cell_mw_cs;
        }   
      },  
      Kokkos::Sum<Real>(total_mass),
      Kokkos::Sum<Real>(mw_density),
      Kokkos::Sum<Real>(mw_velocity),
      Kokkos::Sum<Real>(mw_cs));
```
I get this error:
```
[ 73%] Building CXX object src/CMakeFiles/athenaPK.dir/pgen/cluster/agn_triggering.cpp.o
/u/glines/code/athenapk-project/athenapk/external/parthenon/src/kokkos_abstraction.hpp(628): error: no instance of overloaded function "parthenon::par_dispatch" matches the argument list
            argument types are: (parthenon::LoopPatternMDRange, const char [27], parthenon::DevExecSpace, int, int, int, int, int, int, int, int, lambda [](const int &, const int &, const int &, const int &, parthenon::Real &, parthenon::Real &, parthenon::Real &, parthenon::Real &)->void, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>)
          detected during instantiation of "void parthenon::par_reduce(Args &&...) [with Args=<parthenon::LoopPatternMDRange &, const char (&)[27], parthenon::DevExecSpace, int, int, int &, int &, int &, int &, int &, int &, lambda [](const int &, const int &, const int &, const int &, parthenon::Real &, parthenon::Real &, parthenon::Real &, parthenon::Real &)->void, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>, Kokkos::Sum<parthenon::Real, Kokkos::HostSpace>>]" 
/u/glines/code/athenapk-project/athenapk/src/pgen/cluster/agn_triggering.cpp(214): here

1 error detected in the compilation of "/u/glines/code/athenapk-project/athenapk/src/pgen/cluster/agn_triggering.cpp".
make[2]: *** [src/CMakeFiles/athenaPK.dir/build.make:277: src/CMakeFiles/athenaPK.dir/pgen/cluster/agn_triggering.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1755: src/CMakeFiles/athenaPK.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

pgrete replied:
hmm, likely Parthenon's `par_dispatch` does not support this yet.
Have you tried using a plain `Kokkos::parallel_reduce` ?

forrestglines replied:
We've since moved to `parallel_reduce` and removed the need for this header.

forrestglines commented on tst/regression/test_suites/cluster_agn_triggering/cluster_agn_triggering.py:
```diff
@@ -0,0 +1,439 @@
+        self.uniform_gas_temp = (
+            self.mu * self.m_u / self.k_b * self.uniform_gas_pres / self.uniform_gas_rho
+        )
+        print("Uniform gas temperature: ", self.uniform_gas_temp.in_units("K") * 1.01)
```
Looks like something left over from debugging

forrestglines commented on tst/regression/test_suites/cluster_hydro_agn_feedback/cluster_hydro_agn_feedback.py:
```diff
@@ -0,0 +1,470 @@
+
+        self.norm_tol = 1e-3
+
+        self.steps = 4
+        self.step_params_list = list(
+            itertools.product(
+                ("thermal_only", "kinetic_only", "combined"), (True, False)
+            )
```
Woops, that's definitely wrong.

forrestglines commented on tst/regression/test_suites/cluster_hydro_agn_feedback/cluster_hydro_agn_feedback.py:
```diff
@@ -0,0 +1,470 @@
+                drho_t, dE_t = thermal_feedback(Z, Y, X, dt)
+
+                drho = drho_k + drho_t
+                dMx = dMx_k
+                dMy = dMx_k
+                dMz = dMx_k
```
I got side-tracked while figuring what's wrong with this test. Apparently nothing is injected/changed in either the data read from the AthenaPK output nor the analytical expectation. I don't think the simulation code is broken since I have movies of working kinetic jets, something about the test or comparison is going wrong. I'll have to revisit what exactly is broken in this test later (soon).

forrestglines replied:
Fixed this bug and figured out why the test parameters were injecting too weak a jet to test. Everything is working as intended now.

forrestglines commented on src/pgen/cluster/magnetic_tower.cpp:
```diff
@@ -0,0 +1,314 @@
+        // dE_B = 1/2*( 2*dt*B_old*B_new + dt**2*B_new**2)
+        cons(IEN, k, j, i) += prim(IB1, k, j, i) * b_x + prim(IB2, k, j, i) * b_y +
+                              prim(IB3, k, j, i) * b_z +
+                              0.5 * (b_x * b_x + b_y * b_y + b_z * b_z);
```
As far as I know, this is the only place where I'm adding magnetic energy to a possibly existing magnetic field. The simulation might also start with no existing magnetic field so it's important to include the second order term. The reductions for the magnetic tower also account for the second order term.

forrestglines commented on docs/cluster.md:
```diff
@@ -0,0 +1,396 @@
+The parameters `r_sampling` and `max_dr` determine details about how the
+profile for the initial hydrodynamic equilbrium is integrated. (FIXME -- these
+parameters are esoteric and can potentially be removed. They were added at
+somepoint to ensure that the HSE profile is integrated over a domain that
```
After rereading the code more closely, I don't think `r_sampling` should be removed. It controls the resolution of the 1D profile that we integrate the HSE equations to get density and pressure. We could make it a fixed constant but a runtime parameter would give more control.  I've updated the documentation with a better description of what the parameter does.

forrestglines commented on src/pgen/cluster/hydrostatic_equilibrium_sphere.cpp:
```diff
@@ -50,120 +51,82 @@ HydrostaticEquilibriumSphere<GravitationalField, EntropyProfile>::
+                          "test_he_sphere_r_start_kpc", 1e-3 * units.kpc());
+    const Real test_he_sphere_r_end =
+        pin->GetOrAddReal("problem/cluster/hydrostatic_equilibrium",
+                          "test_he_sphere_r_end_kpc", 4000 * units.kpc());
```
Typos from when the code did have physical units, these variables are actually in code units.

forrestglines commented:
@par-hermes format

forrestglines commented:
@par-hermes format

forrestglines commented:
@par-hermes format

forrestglines commented:
In the current state of the revisions, we have one failing test: `cluster_hydro_agn_feedback`.

The test explores five test cases of thermal only, kinetic only, and combined thermal_kinetic feedback with a tilted and non-tilted kinetic jet. (Only one test for thermal only since the jet tilt is irrelevant). The tests with kinetic feedback are failing in various ways.

1. The change in energy density for all tests with kinetic feedback is off by ~0.15%. This could be related to the question above about injected momentum and change in kinetic energy density.

2. The change in momentum density for the tests with tilted kinetic jets is off by >100%. I suspect there's an issue in the coordinates of precessed jets in either the test or how they are invoked in kinetic jet feedback. The tilted precessed jet coordinates seemed to work correctly in the magnetic tower regression test.

pgrete commented on src/pgen/cluster/jet_coords.hpp:
```diff
@@ -0,0 +1,115 @@
+
+    // Position in jet-cylindrical coordinates
+    r_jet = sqrt(pow(fabs(x_jet), 2) + pow(fabs(y_jet), 2));
+    cos_theta_jet = x_jet / r_jet;
+    sin_theta_jet = y_jet / r_jet;
```
bump

forrestglines replied:
Fixed when `r_jet == 0`

pgrete commented on src/pgen/cluster/magnetic_tower.hpp:
```diff
@@ -0,0 +1,193 @@
+    const parthenon::Real exp_r2_h2 = exp(-pow(r / l_scale_, 2) - pow(h / l_scale_, 2));
+    // Compute the potential in jet cylindrical coordinates
+    a_r = 0.0;
+    a_theta = field_ * l_scale_ * (r / l_scale_) * exp_r2_h2;
```
bump

forrestglines replied:
Only with malicious user input. I've added a `PARTHENON_REQUIRE` for this.

pgrete commented on src/pgen/cluster/magnetic_tower.hpp:
```diff
@@ -0,0 +1,193 @@
+    parthenon::Real r, cos_theta, sin_theta, h;
+    jet_coords_.SimCartToJetCylCoords(x, y, z, r, cos_theta, sin_theta, h);
+
+    return density_ * exp(-(SQR(r) + SQR(h)) / l_mass_scale2_);
```
bump

forrestglines replied:
Only with malicious user input. I've added a `PARTHENON_REQUIRE` for this.

BenWibking commented on src/pgen/cluster/jet_coords.hpp:
```diff
@@ -0,0 +1,115 @@
+        x_sim * sin_theta_jet_axis_ + z_sim * cos_theta_jet_axis_;
+
+    // Position in jet-cylindrical coordinates
+    r_jet = sqrt(pow(fabs(x_jet), 2) + pow(fabs(y_jet), 2));
```
Actually, also at `-O1` and `-O2`. Only at `-O0` do they call `pow`.

forrestglines commented:
@par-hermes format

forrestglines commented on src/pgen/cluster/agn_feedback.cpp:
```diff
@@ -0,0 +1,370 @@
+  }
+  if (pin->DoesParameterExist("problem/cluster/agn_feedback",
+                              "kinetic_jet_temperature")) {
+    kinetic_jet_temperature_ =
+        pin->GetReal("problem/cluster/agn_feedback", "kinetic_jet_temperature");
+
+    const Real He_mass_fraction = pin->GetReal("hydro", "He_mass_fraction");
+    const Real H_mass_fraction = 1.0 - He_mass_fraction;
+    const Real mu = 1 / (He_mass_fraction * 3. / 4. + (1 - He_mass_fraction) * 2);
+    const Real gam = pin->GetReal("hydro", "gamma");
+    const Real gm1 = (gam - 1.0);
+    kinetic_jet_e = units.k_boltzmann() * kinetic_jet_temperature_ /
+                    (mu * units.atomic_mass_unit() * gm1);
+  }
+
+  if (std::isnan(kinetic_jet_velocity_) && std::isnan(kinetic_jet_temperature_)) {
+    // Both velocity and temperature are missing, assume 0K temperature
+    kinetic_jet_velocity_ = units.speed_of_light() * sqrt(2 * (efficiency_));
+    kinetic_jet_temperature_ = 0;
+    kinetic_jet_e = 0;
+    std::cout << "### WARNING Kinetic jet velocity nor temperature not specified. "
+                 "Assuming 0K temperature jet"
+              << std::endl;
+  } else if (std::isnan(kinetic_jet_velocity_)) {
+    // Velocity is missing, compute it from e_jet
+    kinetic_jet_velocity_ = sqrt(2 * (efficiency_ * SQR(units.speed_of_light()) -
+                                      (1.0 - efficiency_) * kinetic_jet_e));
+  } else if (std::isnan(kinetic_jet_temperature_)) {
+    // Temperature is missing, compute e_jet and T_jet from v_jet
+    kinetic_jet_e =
+        (efficiency_ * SQR(units.speed_of_light()) - 0.5 * SQR(kinetic_jet_velocity_)) /
+        (1 - efficiency_);
+    const Real He_mass_fraction = pin->GetReal("hydro", "He_mass_fraction");
+    const Real H_mass_fraction = 1.0 - He_mass_fraction;
+    const Real mu = 1 / (He_mass_fraction * 3. / 4. + (1 - He_mass_fraction) * 2);
+    const Real gam = pin->GetReal("hydro", "gamma");
+    const Real gm1 = (gam - 1.0);
+    kinetic_jet_temperature_ =
+        (kinetic_jet_e * mu * units.atomic_mass_unit() * gm1) / units.k_boltzmann();
```
Reminder to update computing the temperature in this new code that was written about the same time as the mu-update.

forrestglines commented on src/pgen/cluster/jet_coords.hpp:
```diff
@@ -0,0 +1,115 @@
+/************************************************************
+ * Jet Coordinates Factory Class
+ * A factory for creating JetCoords objects given a time
+ * Keeps track of the precession
```
I don't understand some of this question  -- `phi_dot` is constant for a precessing jet.

The way the `JetCoords` class is used is that we create a new one every time we need one. This allow us to execute some of the jet trigonometry functions once at creation rather than at every point. The current implementation is an optimization.

pgrete replied:
Gotcha. Then I misunderstood the comment in the code.
I thought that "keeping track" meant that the object is persistent and updated (according to some prescribed manner so that there's a track record).
This is also why I asked about the varying `phi_dot` as for a constant `phi_dot` things can be calculated at once (as you also say).

forrestglines commented on src/pgen/cluster/magnetic_tower.cpp:
```diff
@@ -0,0 +1,314 @@
+
+        // Add density
+        const Real cell_delta_rho =
+            mt.DensityFromSimCart(coords.Xc<1>(i), coords.Xc<1>(i), coords.Xc<1>(i));
```
Fixed!

forrestglines commented on src/pgen/cluster/magnetic_tower.cpp:
```diff
@@ -0,0 +1,314 @@
+      reducer_sum);
+
+  linear_contrib += mt_power_reduction.data[0];
+  quadratic_contrib += mt_power_reduction.data[1];
```
The magnetic energy injected with a change in magnetic field scales as $`\dot{E} = \alpha  \dot{B} + \beta \dot{B}^2`$, where $`\alpha`$ is the linear contribution and $`\beta`$ is the quadratic contribution, both of which need to be reduced. There's a deeper explanation in the Latex document.

forrestglines commented:
> In the current state of the revisions, we have one failing test: `cluster_hydro_agn_feedback`.
> 
> The test explores five test cases of thermal only, kinetic only, and combined thermal_kinetic feedback with a tilted and non-tilted kinetic jet. (Only one test for thermal only since the jet tilt is irrelevant). The tests with kinetic feedback are failing in various ways.
> 
>     1. The change in energy density for all tests with kinetic feedback is off by ~0.15%. This could be related to the question above about injected momentum and change in kinetic energy density.
> 
>     2. The change in momentum density for the tests with tilted kinetic jets is off by >100%. I suspect there's an issue in the coordinates of precessed jets in either the test or how they are invoked in kinetic jet feedback. The tilted precessed jet coordinates seemed to work correctly in the magnetic tower regression test.

I'd since fixed the precessing jet. It was due to incorrect application of rotations from a small detail in the Sympy functions I use to figure out the coordinate rotation and transformation math. It's working now.

forrestglines commented on src/pgen/cluster/magnetic_tower.cpp:
```diff
@@ -0,0 +1,314 @@
+
+  // Construct magnetic vector potential then compute magnetic fields
+
+  // Currently reallocates this vector potential everytime step and constructs
+  // the potential in a separate kernel. There are two solutions:
+  //  1. Allocate a dependant variable in the hydro package for scratch
+  //  variables, use to store this potential. Would save time in allocations
+  //  but would still require more DRAM memory and two kernel launches
+  //  2. Compute the potential (12 needed in all) in the same kernel,
+  //  constructing the derivative without storing the potential (more
+  //  arithmetically intensive, maybe faster)
```
I've made issue #61 to address this later

forrestglines commented:
@pgrete I've addressed all the comments I can. I've also made issues #46, #61, and #62 to address further changes.

Please take a look at the changes again. Merge if you approve and the CI is passing.

pgrete reviewed (approved):
Final changes look good! Great job, @forrestglines !
Test pass (though I'm not sure why the results are not reported back).
Pulling the trigger.

BenWibking commented:
Did the CI machine die? I don't think it usually takes this long to run.
